addons/resource_distribution_controller.go

```diff
@@ -156,17 +164,32 @@ func (r *ResourceDistributionReconciler) Reconcile(ctx context.Context, req ctrl
 	}
 	if len(templateList.Items) == 0 {
 		logger.Info("No resources to distribute to StorageConsumers.")
-		err = removeFinalizerFromObject(ctx, r.SpokeClient, &addonDeletionlock, ResourceDistributionFinalizer)
-		if err != nil {
+		addonDeletionlock.Data = nil
+		if err = r.SpokeClient.Update(ctx, &addonDeletionlock); err != nil {
 			return ctrl.Result{}, err
 		}
 		return ctrl.Result{}, nil
 	}
 
-	var storageClientMapping *corev1.ConfigMap
-	storageClientMapping, err = utils.GetStorageClientMapping(ctx, r.SpokeClient, r.CurrentNamespace)
-	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("unable to distribute resources to StorageConsumers: %w", err)
+	deletionLockUpdated := false
+	for k := range storageClientMapping.Data {
+		foundK := false
+		for _, tmp := range templateList.Items {
+			if tmp.Labels[utils.CreatedForClientID] == k {
+				foundK = true
+				break
+			}
+		}
+		if !foundK {
+			delete(addonDeletionlock.Data, k)
+			deletionLockUpdated = true
+		}
+	}
+
+	if deletionLockUpdated {
+		if err = r.SpokeClient.Update(ctx, &addonDeletionlock); err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
From what I understood the logic here is something like:
1. Add all clientIDs to the addonDeletionLock and update it
2. If there are no templates make the configmap empty and update it
3. If there are templates but there are not templates for a client then we remove that clientID and update it

The logic could be simplified here to:
1. If there are no templates then update make the map nil as nil, update the configmap and return
2. If there are templates then for each template add the clientID to the configmap and update